packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -6267,46 +6266,77 @@ abstract class RenderObjectElement extends Element {
     return ancestor as RenderObjectElement?;
   }
 
-  ParentDataElement<ParentData>? _findAncestorParentDataElement() {
+  List<ParentDataElement<ParentData>> _findAncestorParentDataElements() {
+    // More than one ParentDataWidget can contribute ParentData, but there are
+    // some constraints.
+    // 1. ParentData can only be written by unique ParentDataWidgets.
+    //    For example, two KeepAlive ParentDataWidgets trying to write to the
+    //    same child is not allowed.
+    // 2. The ParentData itself must be compatible with all ParentDataWidgets
+    //    writing to it.
+    //    For example, TwoDimensionalViewportParentData uses the
+    //    KeepAliveParentDataMixin, so it could be compatible with both
+    //    KeepAlive, and another ParentDataWidget with ParentData type
+    //    TwoDimensionalViewportParentData or a subclass thereof.
+    // The first case is verified here. The second is verified in
+    // debugIsValidRenderObject.
     Element? ancestor = _parent;
-    ParentDataElement<ParentData>? result;
+    final List<ParentDataElement<ParentData>> result = <ParentDataElement<ParentData>>[];
+    final Set<Type> ancestorTypes = <Type>{};
     while (ancestor != null && ancestor is! RenderObjectElement) {
       if (ancestor is ParentDataElement<ParentData>) {
-        result = ancestor;
-        break;
+        ancestorTypes.add(ancestor.runtimeType);
+        result.add(ancestor);
       }
       ancestor = ancestor._parent;
     }
     assert(() {
-      if (result == null || ancestor == null) {
+      if (result.isEmpty || ancestor == null) {
         return true;
       }
-      // Check that no other ParentDataWidgets want to provide parent data.
-      final List<ParentDataElement<ParentData>> badAncestors = <ParentDataElement<ParentData>>[];
-      ancestor = ancestor!._parent;
-      while (ancestor != null && ancestor is! RenderObjectElement) {
-        if (ancestor is ParentDataElement<ParentData>) {
-          badAncestors.add(ancestor! as ParentDataElement<ParentData>);
-        }
-        ancestor = ancestor!._parent;
-      }
-      if (badAncestors.isNotEmpty) {
-        badAncestors.insert(0, result);
+      // Check that no other ParentDataWidgets of the same type want to provide
```

Comment:
Wouldn't we ideally want to assert on the parent data type? If I implement a new `MyPositioned` ParentDataWidget that also operates on `StackParentData` and I mix `MyPositioned` with `Positioned` to write to the same parent data slot I will with the current implementation not get an assertion here, but I am in for a surprise at runtime since `MyPositioned` will silently override the parent data of `Positioned`.

Reply:
Added this as well.